src/java.base/share/classes/java/util/zip/GZIPHeaderBuilder.java

```diff
@@ -0,0 +1,361 @@
+/*
+ * Copyright (C) 2021 THL A29 Limited, a Tencent company. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package java.util.zip;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+/**
+ * This class implements the GZIP file header builder.
+ * The {@code GZIPHeaderData} could be composed by (optionally) call
+ * the method {@code extraFieldBytes}, {@code fileName}, {@code fileComment},
+ * {@code calculateHeaderCRC} and finally call {@code build}.
+ *
+ * @since 18
+ *
+ */
+public class GZIPHeaderBuilder {
+    /* compress method, only {@code Deflater.DEFLATED} supportted. */
+    private byte cm;
+    private byte[] extraFieldBytes;
+    private String filename;
+    private String fileComment;
+    /* Header flags, refer to definition in {@code GZIPHeaderData} */
+    private byte flags;
+    private int headerCrc16;
+
+    /**
+     * Creates a GZIP file header builder.
+     * Current only {@code Deflater.DEFLATED} compress method supportted.
+     *
+     * @since 18
+     */
+    public GZIPHeaderBuilder() {
+        this.cm = (byte)Deflater.DEFLATED;
+        this.extraFieldBytes = null;
+        this.filename = null;
+        this.fileComment = null;
+        this.flags = 0;
+        this.headerCrc16 = 0;
+    }
+
+    /**
+     * Add extra field in GZIP file header.
+     * This method verifies the extra fileds layout per RFC-1952.
+     * See comments of {@code verifyExtraFieldLayout}
```

Comment:
You should specify the layout here than refer to a private method, as private methods don't appear in generated online javadocs